File: 0747. Largest Number At Least Twice of Others/0747.java
// time complexity: O(n) - two simple passes over the array
// space complexity: O(1)
class Solution {
    public int dominantIndex(int[] nums) {
        // find the value and index of the largest element
        int largest = 0, index = 0;
        for (int i=0; i<nums.length; i++) {
            if (nums[i] > largest) {
                largest = nums[i];
                index = i;
            }
        }
        // check that largest is at least twice every other element
        for (int i=0; i<nums.length; i++) {
            if (i != index && largest < nums[i]*2) {
                return -1;
            }
        }
        return index;
    }
}

File: 0747. Largest Number At Least Twice of Others/0747.py
# time complexity: O(n)
# space complexity: O(1)
class Solution(object):
    def dominantIndex(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # initialize top two values and index of largest
        max1, max2, index = -1, -1, -1
        for i, x in enumerate(nums):
            if x > max1:
                max2 = max1  # shift down previous max
                max1 = x
                index = i      # record new max index
            elif x > max2:
                max2 = x     # update second max
        # verify dominance condition
        return -1 if max1 < 2*max2 else index